src/main/java/org/gg/service/UserServiceImpl.java

```
@@ -13,4 +20,40 @@ public class UserServiceImpl implements UserService {
     public UserServiceImpl(UserRepository userRepository) {
         this.userRepository = userRepository;
     }
+
+    @Override
+    public User addUser(User user) {
+        user.setSalt(HashUtil.generateSalt());
+        String password = user.getPassword();
+        try {
+            user.setPassword(HashUtil.hashPassword(password, user.getSalt()));
+        } catch (NoSuchAlgorithmException e) {
+            throw new RuntimeException(e);
+        }
+        return userRepository.save(user);
+    }
+
+    @Override
+    public Optional<User> getUserById(String id) {return userRepository.findById(id);}
+
+
+    @Override
+    public User updateUser(String id, User updatedUser) {
+        Optional<User> optionalExistingUser = userRepository.findById(id);
+
+        if(optionalExistingUser.isPresent()){
+            User existingUser = optionalExistingUser.get();
+        //spring framework method which copies values from one object to another by providing a list of null values to populate
+            BeanUtils.copyProperties(updatedUser, existingUser, BeanUtil.getNullPropertyNames(updatedUser));
+            existingUser.setId(id);
+
+            return userRepository.save(existingUser);
+        }
+        return null;
```

Should we return an error response if no user if found to update?

---

Hi Chris, thank you for great feedback. Regarding your comment above, there are two possible scenarios that can happen in that method inside the UserServiceImpl. If a user exists, the user will get updated, otherwise, it will return user = null. 

Having a look at the UserController method updateUser, this is handled by marking the user as optional using the method findUserById, which implicitly, will return a status code of 404. The method in UserServiceImpl will not be reached to throw an exception.